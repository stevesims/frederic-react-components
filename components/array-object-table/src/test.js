import React from 'react';
import { mount, shallow } from 'enzyme';

import ArrayObjectTable from '.';

const fields = [
  { key: 'one', heading: 'One' },
  { key: 'two', heading: 'Two', transform: () => 'two' },
  { key: 'three', heading: 'Three' },
];

const array = [
  {},
  { one: 'test', two: 'test' },
];

const getRows = wrapper => wrapper.find('Table').prop('rows');

describe('ArrayObjectTable', () => {
  it('renders with required props', () => {
    const wrapper = shallow(<ArrayObjectTable fields={fields} array={array} />);
    expect(wrapper.exists()).toBe(true);
  });

  describe('renders table data', () => {
    it('renders expected table heading cells ', () => {
      const wrapper = mount(<ArrayObjectTable fields={fields} array={array} />);
      const titles = wrapper.find('Table').prop('titles');
      expect(titles).toEqual(['One', 'Two', 'Three']);
    });

    it('renders expected table data cells', () => {
      const wrapper = mount(<ArrayObjectTable fields={fields} array={array} />);
      expect(getRows(wrapper)).toEqual([['-', 'two', '-'], ['test', 'two', '-']]);
    });
  });

  describe('responds as expected to additional props', () => {
    it('renders a table heading if provided', () => {
      const wrapper = shallow(<ArrayObjectTable fields={fields} array={array} title="Heading"/>);
      expect(wrapper.contains('Heading')).toBe(true);
    });

    it('skips empty rows if skipEmptyRows is true', () => {
      const wrapper = mount(<ArrayObjectTable fields={fields} array={array} skipEmptyRows/>);
      expect(getRows(wrapper)).toEqual([['test', 'two', '-']]);
    });

    it('renders nothing if no rows are returned and hideWithNoValues is true', () => {
      const emptyArray = [{}, {}, {}];
      const wrapper = shallow(<ArrayObjectTable fields={fields} array={emptyArray} skipEmptyRows hideWithNoValues />);
      expect(wrapper.html()).toBe(null);
    });

    it('renders one tbody row if no rows are returned and hideWithNoValues is not true', () => {
      const emptyArray = [{}, {}, {}];
      const wrapper = mount(<ArrayObjectTable fields={fields} array={emptyArray} skipEmptyRows />);
      expect(wrapper.find('tbody tr').length).toBe(1);
    });
  });

  describe('uses transforms as expected', () => {
    const fieldsWithTransforms = [
      { key: 'one', heading: 'One' },
      { key: 'two', heading: 'Two', transform: value => value ? value.toLowerCase() : '-' },
      { key: 'three', heading: 'Three' },
      { key: 'four', heading: 'Four', transform: value => value ? value.toLowerCase() : '*' },
    ];
    const arrayOfOnlyTwo = [
      {one: 'One', two: 'Two'},
    ];

    it('sends table row cells as expected', () => {
      const wrapper = mount(<ArrayObjectTable
        fields={fieldsWithTransforms}
        array={arrayOfOnlyTwo}
        title="Heading"
        defaultTransform={value => (value || 'x')}
      />);
      expect(getRows(wrapper)).toEqual([['One', 'two', 'x', '*']]);
    });
  });

  it('matches snapshot', () => {
    const wrapper = mount(<ArrayObjectTable fields={fields} array={array} skipEmptyRows />);
    expect(wrapper).toMatchSnapshot();
  });
});
