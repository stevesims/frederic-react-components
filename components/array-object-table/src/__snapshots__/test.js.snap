// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayObjectTable matches snapshot 1`] = `
.emotion-14 {
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-14 {
    margin-bottom: 15px;
  }
}

.emotion-12 {
  display: table;
  font-family: "nta",Arial,sans-serif;
  width: 100%;
  table-layout: fixed;
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-12 {
    margin-bottom: 15px;
  }
}

.emotion-0 {
  border-bottom: 1px solid #bfc1c3;
  display: table-cell;
  font-size: 14px;
  padding: 15px 4px;
  vertical-align: top;
  font-weight: bold;
  text-align: left;
}

.emotion-0:first-child {
  padding: 15px 4px 15px 0;
}

.emotion-0:last-child {
  padding: 15px 0 15px 4px;
}

.emotion-6 {
  border-bottom: 1px solid #bfc1c3;
  display: table-cell;
  font-size: 14px;
  padding: 15px 4px;
  vertical-align: top;
}

.emotion-6:first-child {
  padding: 15px 4px 15px 0;
}

.emotion-6:last-child {
  padding: 15px 0 15px 4px;
}

<ArrayObjectTable
  array={
    Array [
      Object {},
      Object {
        "one": "test",
        "two": "test",
      },
    ]
  }
  defaultTransform={[Function]}
  fields={
    Array [
      Object {
        "heading": "One",
        "key": "one",
      },
      Object {
        "heading": "Two",
        "key": "two",
        "transform": [Function],
      },
      Object {
        "heading": "Three",
        "key": "three",
      },
    ]
  }
  hideWithNoValues={false}
  skipEmptyRows={true}
  title={null}
>
  <Styled(Table)
    flexibleColumns={false}
    nameByRow={false}
    names={Array []}
    rowIncludesHeading={false}
    rows={
      Array [
        Array [
          "test",
          "two",
          "-",
        ],
      ]
    }
    titles={
      Array [
        "One",
        "Two",
        "Three",
      ]
    }
  >
    <Table
      className="emotion-14"
      flexibleColumns={false}
      nameByRow={false}
      names={Array []}
      rowIncludesHeading={false}
      rows={
        Array [
          Array [
            "test",
            "two",
            "-",
          ],
        ]
      }
      titles={
        Array [
          "One",
          "Two",
          "Three",
        ]
      }
    >
      <TableContainer
        className="emotion-14"
        flexibleColumns={false}
      >
        <table
          className="emotion-12 emotion-13"
        >
          <thead>
            <tr>
              <TableHeading
                key="0"
              >
                <th
                  className="emotion-0 emotion-1"
                >
                  One
                </th>
              </TableHeading>
              <TableHeading
                key="1"
              >
                <th
                  className="emotion-0 emotion-1"
                >
                  Two
                </th>
              </TableHeading>
              <TableHeading
                key="2"
              >
                <th
                  className="emotion-0 emotion-1"
                >
                  Three
                </th>
              </TableHeading>
            </tr>
          </thead>
          <tbody>
            <tr
              key="0"
            >
              <TableData
                key="0"
              >
                <td
                  className="emotion-6 emotion-7"
                >
                  test
                </td>
              </TableData>
              <TableData
                key="1"
              >
                <td
                  className="emotion-6 emotion-7"
                >
                  two
                </td>
              </TableData>
              <TableData
                key="2"
              >
                <td
                  className="emotion-6 emotion-7"
                >
                  -
                </td>
              </TableData>
            </tr>
          </tbody>
        </table>
      </TableContainer>
    </Table>
  </Styled(Table)>
</ArrayObjectTable>
`;
